keps/sig-api-machinery/20190607-owner-ref.md

```diff
@@ -0,0 +1,127 @@
+---
+title: OwnerReference Resource Field
+authors:
+  - "@deads2k"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-api-machinery
+reviewers:
+  - "@sttts"
+  - "@caesarxuchao"
+  - "@liggitt"
+approvers:
+  - "@caesarxuchao"
+  - "@lavalamp"
+editor:
+  name: "@deads2k"
+creation-date: 2019-06-07
+last-updated: 2019-06-07
+status: implementable
+---
+
+# OwnerReference Resource Field
+
+## Table of Contents
+
+## Summary
+
+OwnerReferences are used to track garbage collection references.  Today they use `kind`.  This is a serialization format
+not a URL to access on the kube-apiserver.  To find a URL, a series of upwards of 20 discovery calls are made to find
+every resource in the cluster to find a list of potential resources (URLs more or less).  This approximates user intent
+and mostly works, but it is inefficient and unnecessarily vulnerable to network outages.
+
+This KEP proposes adding an optional `resource` field to OwnerReferences.  If provided, this will be the authoritative value and
+no lookup or approximation of user intent will be required.  If not provided, the `kind` will be used as it is today.
+If both are provided, old clusters continue to work and new clusters will be more efficient and resilient.
+ 
+```yaml
+ownerReferences:
+- apiVersion: apps/v1
+  kind: DaemonSet
+  resource: daemonsets
+```
+
+## Motivation
+
+* Allow precision.  `kind` is a on to many relationship that makes the exact object reference inefficient or impossible to determine.
+* Remove remote calls.  Because a `kind` to `resource` mapping requires discovery, it is vulnerable to network segmentation
+ 
+### Goals
+
+* Improve reliability. 
+* Improve efficiency.
+* Improve predictability.
+* Avoid disrupting existing clients.
+
+### Non-Goals
+
+* Improving discovery.  No amount of improving speed can fix predictability and we're already doing what caching is practical.
+* Improve rest mapping.  No amount of feature enhancement can fix predictability.
+
+## Proposal
+
+Add a new, optional field to `OwnerReferences` that represents the `resource`.
+
+```go
+type OwnerReference struct {
+	// Kind of the referent.
+	Kind string `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+	// Resource of the referent.
+	// +optional
+	Resource string `json:"resource,omitempty" protobuf:"bytes,8,opt,name=resource"`
+}
+```
+
+The only impact is in the case of a new cluster with both fields set.  In this case, the new cluster does not have to 
+perform any lookups and will trust the client's specification of `resource`.
+
+Existing clients are not required to make this change, but it will improve reliability.  If a client sets the value improperly,
+then the behavior is congruent to behavior with a mis-set `kind`.  Generally speaking, GC collects the object immediately.  
```

Comment:
nit: Today, GC will [record the error](https://github.com/kubernetes/kubernetes/blob/78220fe3803c056b89bbed84a8d4512a4af9870b/pkg/controller/garbagecollector/garbagecollector.go#L339) and retry.

The behavior with the mis-set `resource` is still congruent to a mis-set `kind`.